src/storage/FirebaseStorage.js

```diff
@@ -0,0 +1,97 @@
+(function() {
+  "use strict";
+
+  function unwrapStorageSnapshot(storageSnapshot) {
+    return {
+      bytesTransferred: storageSnapshot.bytesTransferred,
+      downloadURL: storageSnapshot.downloadURL,
+      metadata: storageSnapshot.metadata,
+      ref: storageSnapshot.ref,
+      state: storageSnapshot.state,
+      task: storageSnapshot.task,
+      totalBytes: storageSnapshot.totalBytes
+    };
+  }
+
+  function _$put(storageRef, file, $digestFn) {
+    var task = storageRef.put(file);
+
+    return {
+      $progress: function $progress(callback) {
+        task.on('state_changed', function () {
+          $digestFn(function () {
+            callback([unwrapStorageSnapshot(task.snapshot)]);
+          });
+        });
+      },
+      $error: function $error(callback) {
+        task.on('state_changed', function () {}, function (err) {
+          $digestFn(function () {
+            callback([err]);
+          });
+        });
+      },
+      $complete: function $complete(callback) {
+        task.on('state_changed', function () {}, function () {}, function () {
+          $digestFn(function () {
+            callback([unwrapStorageSnapshot(task.snapshot)]);
+          });
+        });
+      },
+      $cancel: task.cancel,
+      $resume: task.resume,
+      $pause: task.pause,
+      then: task.then,
+      catch: task.catch,
+      _task: task
+    };
+  }
+
+  function isStorageRef(value) {
+    value = value || {};
+    return typeof value.put === 'function';
+  }
+
+  function _assertStorageRef(storageRef) {
+    if (!isStorageRef(storageRef)) {
+      throw new Error('$firebaseStorage expects a storage reference from firebase.storage().ref()');
```

Comment:
Suggestion: "$firebaseStorage expects a Storage reference"
* Capitalizing "Storage"
* Taking out "from firebase.storage().ref()" since it could also come from "app.storage().ref()"